File: backend/src/main/java/be/lutske/leolegacy/interfaceadapter/rest/CategoryResource.java
package be.lutske.leolegacy.interfaceadapter.rest;

import be.lutske.leolegacy.application.usecase.ListCategoriesUseCase;
import be.lutske.leolegacy.domain.category.Category;
import be.lutske.leolegacy.interfaceadapter.dto.CategoryResponse;
import jakarta.inject.Inject;
import jakarta.ws.rs.*;
import jakarta.ws.rs.core.MediaType;
import java.util.List;

@Path("/api/categories")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
public class CategoryResource {

    @Inject
    ListCategoriesUseCase listCategoriesUseCase;

    @GET
    public Response listCategories() {
        List<Category> categories = listCategoriesUseCase.execute();
        return Response.ok(categories.stream()
                .map(CategoryResponse::fromDomain)
                .toList())
                .build();
    }
}

File: backend/src/main/kotlin/be/lutske/leolegacy/interfaceadapter/rest/CategoryResource.kt
package be.lutske.leolegacy.interfaceadapter.rest

import be.lutske.leolegacy.infrastructure.persistence.repository.CategoryRepository
import be.lutske.leolegacy.infrastructure.persistence.repository.RecipeRepository
import jakarta.ws.rs.GET
import jakarta.ws.rs.Path
import jakarta.ws.rs.Produces
import jakarta.ws.rs.core.MediaType

/**
 * REST resource for recipe categories.
 */
@Path("/api/categories")
class CategoryResource(
    private val categoryRepository: CategoryRepository,
    private val recipeRepository: RecipeRepository
) {

    @GET
    @Produces(MediaType.APPLICATION_JSON)
    fun listCategories(): List<CategoryResponse> {
        return categoryRepository.findAllOrderedByName().map { cat ->
            val count = recipeRepository.findByCategoryId(cat.id).size.toLong()
            CategoryResponse(
                id = cat.id,
                name = cat.name,
                recipeCount = count
            )
        }
    }
}
